src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,532 @@
+  refine rec_on_mul h0 _ (λ p pp, _) (λ m n hm hn, _) b,
+  { simp only [jacobi_sym, factors_one, list.prod_nil, one_mul, list.pmap], },
+  { simp_rw [← @legendre_sym.to_jacobi_sym p ⟨pp⟩, @legendre_sym_mul p ⟨pp⟩], },
+  { by_cases hmz : m = 0,
+    { rw [hmz, zero_mul], exact h0, },
+    by_cases hnz : n = 0,
+    { rw [hnz, mul_zero], exact h0, },
+    simp_rw [@jacobi_sym_mul_right _ _ _ ⟨hmz⟩ ⟨hnz⟩],
+    rw [hm, hn, mul_mul_mul_comm], },
+end
+
+/-- We have that `(a^e / b) = (a / b)^e` for the Jacobi symbol. -/
+lemma jacobi_sym_pow_left (a : ℤ) (e b : ℕ) : [a ^ e | b]ⱼ = [a | b]ⱼ ^ e :=
+begin
+  induction e with e ih,
+  { rw [pow_zero, pow_zero, jacobi_sym_one_left], },
+  { rw [pow_succ, pow_succ, jacobi_sym_mul_left, ih], }
+end
+
+/-- We have that `(a / b^e) = (a / b)^e` for the Jacobi symbol. -/
+lemma jacobi_sym_pow_right (a : ℤ) (b e : ℕ) : [a | b ^ e]ⱼ = [a | b]ⱼ ^ e :=
+begin
+  induction e with e ih,
+  { rw [pow_zero, pow_zero, jacobi_sym_one_right], },
+  { by_cases hb : b = 0,
+    { rw [hb, zero_pow (succ_pos e), jacobi_sym_zero_right, one_pow], },
+    { haveI : ne_zero b := ⟨hb⟩,
+      haveI : ne_zero (b ^ e) := ⟨pow_ne_zero e hb⟩,
+      rw [pow_succ, pow_succ, jacobi_sym_mul_right, ih], } }
+end
+
+/-- The Jacobi symbol `(a / b)` takes the value `1` or `-1` if `a` and `b` are coprime. -/
+lemma jacobi_sym_eq_one_or_neg_one {a : ℤ} {b : ℕ} (h : a.gcd b = 1) :
+  [a | b]ⱼ = 1 ∨ [a | b]ⱼ = -1 :=
+begin
+  refine rec_on_mul (λ _, or.inl $ jacobi_sym_zero_right a)
+          (λ _, or.inl $ jacobi_sym_one_right a) (λ p pp hpg, _) (λ m n hm hn hmng, _) b h,
+  { simp_rw [← @legendre_sym.to_jacobi_sym p ⟨pp⟩],
+    exact @legendre_sym_eq_one_or_neg_one p ⟨pp⟩ _ (ne_zero_of_gcd_eq_one pp hpg), },
+  { by_cases hm0 : m = 0,
+    { rw [hm0, zero_mul],
+      exact or.inl (jacobi_sym_zero_right a), },
+    by_cases hn0 : n = 0,
+    { rw [hn0, mul_zero],
+      exact or.inl (jacobi_sym_zero_right a), },
+    rw [nat.cast_mul] at hmng,
+    have hng := hn (int.gcd_eq_one_of_gcd_mul_eq_one_right hmng),
+    simp_rw [@jacobi_sym_mul_right _ _ _ ⟨hm0⟩ ⟨hn0⟩],
+    cases hm (int.gcd_eq_one_of_gcd_mul_eq_one_left hmng) with hl hr,
+    { rwa [hl, one_mul], },
+    { rw [hr, neg_mul, one_mul, neg_inj, neg_eq_iff_neg_eq],
+      exact or.dcases_on hng or.inr (λ hr', or.inl hr'.symm), } },
+end
+
+/-- The square of the Jacobi symbol `(a / b)` is `1` when `a` and `b` are coprime. -/
+lemma jacobi_sym_sq_one {a : ℤ} {b : ℕ} (h : a.gcd b = 1) : [a | b]ⱼ ^ 2 = 1 :=
+by cases jacobi_sym_eq_one_or_neg_one h with h₁ h₁; rw h₁; refl
+
+/-- The Jacobi symbol `(a^2 / b)` is `1` when `a` and `b` are coprime. -/
+lemma jacobi_sym_sq_one' {a : ℤ} {b : ℕ} (h : a.gcd b = 1) : [a ^ 2 | b]ⱼ = 1 :=
+by rw [pow_two, jacobi_sym_mul_left, ← pow_two, jacobi_sym_sq_one h]
+
+/-- The Jacobi symbol `(a / b)` depends only on `a` mod `b`. -/
+lemma jacobi_sym_mod_left (a : ℤ) (b : ℕ) : [a | b]ⱼ = [a % b | b]ⱼ :=
```

Comment:
An easier proof would start with `congr_arg list.prod _ . $ list.pmap_congr _ begin sorry, end` but it requires merging master to get #16220.

Reply:
OK.